packages/jupyterlab-manager/src/manager.ts

```diff
@@ -189,29 +201,53 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
       return;
     }
     await this.context.session.ready;
+    if ((this.context.session.kernel as any).handleComms === false) {
+      return;
+    }
     const comm_ids = await this._get_comm_info();
 
-    // For each comm id, create the comm, and request the state.
+    // For each comm id that we do not know about, create the comm, and request the state.
     const widgets_info = await Promise.all(Object.keys(comm_ids).map(async (comm_id) => {
-      const comm = await this._create_comm(this.comm_target_name, comm_id);
-      const update_promise = new Promise<Private.ICommUpdateData>((resolve, reject) => {
-        comm.on_msg((msg) => {
-          put_buffers(msg.content.data.state, msg.content.data.buffer_paths, msg.buffers);
-          // A suspected response was received, check to see if
-          // it's a state update. If so, resolve.
-          if (msg.content.data.method === 'update') {
-            resolve({
-              comm: comm,
-              msg: msg
+      try {
+        await this.get_model(comm_id);
+        // If we successfully get the model, do no more.
+        return;
+      } catch (e) {
+        // If we have the widget model not found error, then we can create the
+        // widget. Otherwise, rethrow the error. We have to check the error
+        // message text explicitly because the get_model function in this
+        // class throws a generic error with this specific text.
+        if (e.message !== 'widget model not found') {
+          throw e;
+        }
+        const comm = await this._create_comm(this.comm_target_name, comm_id);
+
+        let msg_id: string;
+        const widget_info = new PromiseDelegate<Private.ICommUpdateData>();
```

Comment:
It's a little confusing that this is called `widget_info` vs the outside `widgets_info` (I wasted a few cycles tracing this).

Reply:
widget_info is a single thing in widgets_info list - so it makes sense to me. However, I agree it could be easy to miss the `s` in the middle of the word when reading the code (as opposed to an s at the end).